apollo-router/src/plugins/diagnostics/memory/mod.rs

```diff
@@ -0,0 +1,386 @@
+        Self { output_directory }
+    }
+
+    /// Handle GET /diagnostics/memory/status
+    pub(super) async fn handle_status(&self, request: Request) -> Result<Response, BoxError> {
+        // On Linux, use jemalloc APIs
+        #[cfg(target_os = "linux")]
+        let status = {
+            tokio::task::spawn_blocking(move || -> Result<serde_json::Value, MemoryError> {
+                // Read profiling status from jemalloc
+                let profiling_active =
+                    unsafe { tikv_jemalloc_ctl::raw::read::<bool>(b"prof.active\0") }
+                        .map_err(|e| MemoryError::JemallocControl(e.to_string()))?;
+
+                Ok(json!({
+                    "profiling_active": profiling_active,
+                    "status": if profiling_active { "active" } else { "inactive" },
+                    "platform": "linux",
+                    "heap_dumps_available": true
+                }))
+            })
+            .await
+            .map_err(|e| MemoryError::TaskFailed(format!("Task failed: {}", e)))
+            .and_then(|r| r)
+            .map_err(|e| BoxError::from(e.to_string()))?
+        };
+
+        // On non-Linux platforms, return basic status
+        #[cfg(not(target_os = "linux"))]
+        let status = json!({
+            "profiling_active": false,
+            "status": "not_available",
+            "platform": std::env::consts::OS,
+            "heap_dumps_available": false,
+            "message": "Memory profiling requires Linux platform with jemalloc"
+        });
+
+        Response::http_response_builder()
+            .response(
+                http::Response::builder()
+                    .status(StatusCode::OK)
+                    .header("content-type", "application/json")
+                    .body(body::from_bytes(serde_json::to_vec(&status)?))?,
+            )
+            .context(request.context)
+            .build()
+    }
+
+    /// Handle POST /diagnostics/memory/start
+    pub(super) async fn handle_start(&self, request: Request) -> Result<Response, BoxError> {
+        tracing::info!("Memory profiling start requested");
+
+        // On Linux, use jemalloc APIs
+        #[cfg(target_os = "linux")]
+        let (status_code, response) = {
+            match tokio::task::spawn_blocking(move || -> Result<(), MemoryError> {
+                // Enable jemalloc profiling
+                unsafe { tikv_jemalloc_ctl::raw::write::<bool>(b"prof.active\0", true) }
+                    .map_err(|e| MemoryError::JemallocControl(e.to_string()))?;
+
+                // Verify it was enabled
+                let active = unsafe { tikv_jemalloc_ctl::raw::read::<bool>(b"prof.active\0") }
+                    .map_err(|e| MemoryError::JemallocControl(e.to_string()))?;
+
+                if !active {
+                    return Err(MemoryError::JemallocControl(
+                        "Failed to activate profiling".to_string(),
+                    ));
+                }
```

Comment:
Do we know what might cause this to happen?

Reply:
I think we may be able to remove the check, but I'm not 100% sure. The error handling for jemalloc was difficult to figure out. When you don't have the right env variables set then I think it may silently ignore the call to activate profiling.